File: src/voidx/tools/repomap.py
"""Repo map tool — structural overview of the codebase for LLM context.

Extracts function/class signatures from source files using regex-based
parsing. Groups by file, limits output to a token budget (~4000 tokens).

Supports Python by default. Extensible to other languages.
"""

from __future__ import annotations

import re
from pathlib import Path

from pydantic import BaseModel, Field

from voidx.tools.base import BaseTool, model_to_json_schema, ToolContext, ToolResult, resolve_safe

OUTPUT_TOKEN_BUDGET = 4000
DIR_ENTRY_LIMIT = 200

SKIP_DIRS = {
    ".git", ".venv", "venv", "node_modules", "__pycache__",
    ".pytest_cache", ".mypy_cache", ".tox", ".eggs",
    ".idea", ".vscode", "dist", "build", "opencode",
    ".claude", ".ruff_cache",
}
SKIP_SUFFIXES = {
    ".pyc", ".pyo", ".so", ".dll", ".exe", ".bin",
    ".png", ".jpg", ".jpeg", ".gif", ".ico", ".pdf",
    ".zip", ".tar", ".gz", ".whl", ".egg",
}


class RepoMapInput(BaseModel):
    path: str | None = Field(
        default=None,
        description="Subdirectory to focus on. Defaults to workspace root."
    )
    detail: str = Field(
        default="overview",
        description="'overview' = top-level symbols only, 'signatures' = all function/class signatures"
    )
    pattern: str | None = Field(
        default=None,
        description="Glob pattern to filter files, e.g. '*.py' or 'src/**/*.ts'"
    )


class RepoMapTool(BaseTool):
    id = "repo_map"
    description = (
        "Structural map of the codebase: file tree with function/class signatures. "
        "detail='overview' returns top-level symbols only. detail='signatures' "
        "returns all symbols including methods. Narrow with path or pattern."
    )

    def parameters_schema(self) -> dict:
        return model_to_json_schema(RepoMapInput)

    async def execute(self, args: dict, ctx: ToolContext) -> ToolResult:
        inp = RepoMapInput.model_validate(args)
        base = Path(ctx.workspace)
        if inp.path:
            root = resolve_safe(ctx.workspace, inp.path)
            if root is None:
                return ToolResult(output=f"Path traversal blocked: {inp.path}")
        else:
            root = base

        if not root.exists():
            return ToolResult(output=f"Path not found: {root}")

        files = _collect_files(root, inp.pattern)
        if not files:
            return ToolResult(output=f"No source files found in: {root}")

        detail = inp.detail
        entries: list[str] = []
        total_tokens = 0

        for f in files:
            rel = str(f.relative_to(base)).replace("\\", "/")

            if detail == "overview":
                symbols = _extract_top_level(f)
            else:
                symbols = _extract_all_symbols(f)

            if symbols:
                entry = f"{rel}:\n" + "".join(symbols)
            else:
                entry = f"{rel}: (no symbols)  " if detail == "signatures" else ""

            est = _estimate_tokens(entry)
            if total_tokens + est > OUTPUT_TOKEN_BUDGET:
                remaining = len(files) - len(entries)
                entries.append(f"\n... ({remaining} more files omitted — token limit)")
                break

            if entry:
                entries.append(entry)
                total_tokens += est

        if not entries:
            return ToolResult(output=f"No recognizable symbols found in: {root}")

        return ToolResult(
            title=f"Repo map: {root.relative_to(base)} ({len(entries)} files)",
            output="\n".join(entries),
            metadata={
                "root": str(root.relative_to(base)),
                "files": len(entries),
                "detail": detail,
            },
        )


# ── file collection ──────────────────────────────────────────────────────────

def _collect_files(root: Path, pattern: str | None) -> list[Path]:
    files: list[Path] = []

    def _scan(dir_path: Path):
        try:
            entries = sorted(dir_path.iterdir(), key=lambda p: (not p.is_dir(), p.name.lower()))
        except PermissionError:
            return
        for entry in entries:
            if entry.name in SKIP_DIRS or entry.name.startswith("."):
                continue
            if entry.is_dir():
                _scan(entry)
            elif entry.is_file():
                if entry.suffix in SKIP_SUFFIXES:
                    continue
                if pattern and not entry.match(pattern):
                    continue
                if _is_source(entry):
                    files.append(entry)

    _scan(root)
    return files


def _is_source(path: Path) -> bool:
    ext_map = {
        ".py": True, ".ts": True, ".tsx": True, ".js": True, ".jsx": True,
        ".go": True, ".rs": True, ".java": True, ".kt": True, ".swift": True,
        ".c": True, ".h": True, ".cpp": True, ".hpp": True, ".rb": True,
        ".php": True, ".css": True, ".scss": True, ".less": True,
        ".sql": True, ".sh": True, ".bash": True, ".zsh": True,
        ".yaml": True, ".yml": True, ".toml": True, ".json": True,
        ".md": True, ".txt": True, ".cfg": True, ".ini": True,
        ".dockerfile": True, ".makefile": True, ".mk": True,
    }
    name = path.name.lower()
    if name == "dockerfile" or name == "makefile":
        return True
    return ext_map.get(path.suffix, False)


# ── Python symbol extraction ─────────────────────────────────────────────────


def _extract_all_symbols(f: Path) -> list[str]:
    if f.suffix == ".py":
        return _extract_python_symbols(f)
    return _extract_generic(f)


def _extract_top_level(f: Path) -> list[str]:
    if f.suffix == ".py":
        return _extract_python_symbols(f, top_level_only=True)
    return _extract_generic(f)


def _extract_python_symbols(f: Path, top_level_only: bool = False) -> list[str]:
    try:
        text = f.read_text(encoding="utf-8", errors="replace")
    except Exception:
        return []

    lines: list[str] = []
    decorator: str | None = None

    _RE_COMBINED = re.compile(
        r"^(\s*)class\s+(\w+)\s*(?:\(([^)]*?)\))?\s*:"
        r"|^(\s*)(?:async\s+)?def\s+(\w+)\s*\(([^)]*)\)\s*(?:->[^:]+?)?\s*:"
        r"|^[ \t]*@(\w+)",
        re.MULTILINE,
    )

    for match in _RE_COMBINED.finditer(text):
        indent = match.group(1) or match.group(4) or ""
        level = len(indent) // 4 if indent else 0

        if match.group(0).startswith("@"):
            decorator = match.group(7) or ""
            continue

        is_class = match.group(2) is not None
        func_name = match.group(5)
        name = match.group(2) or func_name
        args = match.group(3) or match.group(6) or ""

        if top_level_only and level > 0:
            continue

        prefix = "  " * level
        if is_class:
            tag = ""
            if decorator:
                tag = f"@{decorator}\n{prefix}"
                decorator = None
            base = f"({match.group(3)})" if match.group(3) else ""
            lines.append(f"{prefix}{tag}class {name}{base}\n")
        else:
            tag = ""
            if decorator:
                tag = f"@{decorator} "
                decorator = None
            if match.group(0).lstrip().startswith("async"):
                tag = "async "
            lines.append(f"{prefix}{tag}def {name}({args})\n")

    return lines


def _extract_generic(f: Path) -> list[str]:
    """Minimal extraction for non-Python files: just report the file."""
    return []


# ── token estimation ─────────────────────────────────────────────────────────

def _estimate_tokens(text: str) -> int:
    """Quick token estimate: ~4 chars per token for code."""
    return max(1, len(text) // 4)

File: crates/voidx-tools/src/repomap.rs
//! RepoMap tool — structural map of the codebase.
//!
//! Ported from `src/voidx/tools/repomap.py`.

use crate::base::{Tool, ToolContext, ToolResult};
use crate::error::ToolError;
use async_trait::async_trait;
use std::path::Path;

pub struct RepoMapTool;

#[async_trait]
impl Tool for RepoMapTool {
    fn id(&self) -> &'static str {
        "repo_map"
    }

    fn description(&self) -> &'static str {
        "Structural map of the codebase: file tree with function/class signatures. Use 'detail=overview' for top-level symbols only, 'detail=signatures' for all symbols including methods. Narrow with path or pattern."
    }

    fn parameters_schema(&self) -> serde_json::Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "path": {
                    "type": "string",
                    "description": "Subdirectory to focus on. Defaults to workspace root."
                },
                "pattern": {
                    "type": "string",
                    "description": "Glob pattern to filter files, e.g. '*.py' or 'src/**/*.ts'"
                },
                "detail": {
                    "type": "string",
                    "enum": ["overview", "signatures"],
                    "description": "'overview' = top-level symbols only, 'signatures' = all function/class signatures. Default: 'overview'"
                }
            }
        })
    }

    async fn execute(
        &self,
        args: serde_json::Value,
        ctx: &ToolContext,
    ) -> Result<ToolResult, ToolError> {
        let path = args.get("path").and_then(|v| v.as_str()).unwrap_or(".");
        let pattern = args.get("pattern").and_then(|v| v.as_str()).unwrap_or("**/*");
        let detail = args.get("detail").and_then(|v| v.as_str()).unwrap_or("overview");

        let workspace = &ctx.workspace;
        let target = workspace.join(path);

        if !target.exists() {
            return Ok(ToolResult::new(format!("Path not found: {}", target.display())));
        }

        let mut output = String::new();
        output.push_str(&format!("Codebase map: {} (detail={})\n\n", target.display(), detail));

        // Walk the directory tree
        if let Ok(entries) = walk_dir(&target, pattern, detail) {
            output.push_str(&entries);
        } else {
            output.push_str("(could not read directory)");
        }

        Ok(ToolResult::new(output))
    }
}

fn walk_dir(root: &Path, pattern: &str, detail: &str) -> Result<String, std::io::Error> {
    let mut output = String::new();
    let mut file_count = 0u32;
    let mut dir_count = 0u32;

    fn walk(
        dir: &Path,
        root: &Path,
        pattern: &str,
        detail: &str,
        output: &mut String,
        file_count: &mut u32,
        dir_count: &mut u32,
        depth: usize,
    ) -> Result<(), std::io::Error> {
        let entries = std::fs::read_dir(dir)?;
        let mut entries: Vec<_> = entries.filter_map(|e| e.ok()).collect();
        entries.sort_by_key(|e| {
            let is_dir = e.file_type().map(|t| t.is_dir()).unwrap_or(false);
            (!is_dir, e.file_name())
        });

        for entry in entries {
            let name = entry.file_name().to_string_lossy().to_string();
            // Skip hidden and build directories
            if name.starts_with('.') || name == "node_modules" || name == "target"
                || name == "__pycache__" || name == ".git" || name == "build"
                || name == "dist" || name == ".venv"
            {
                continue;
            }

            let path = entry.path();
            let indent = "  ".repeat(depth);

            if entry.file_type()?.is_dir() {
                *dir_count += 1;
                output.push_str(&format!("{}{}/\n", indent, name));
                walk(&path, root, pattern, detail, output, file_count, dir_count, depth + 1)?;
            } else {
                // Check pattern match
                let rel = path.strip_prefix(root).unwrap_or(&path);
                if matches_pattern(rel.to_string_lossy().as_ref(), pattern) {
                    *file_count += 1;
                    if detail == "signatures" {
                        // Try to extract signatures for code files
                        let sigs = extract_signatures(&path);
                        if sigs.is_empty() {
                            output.push_str(&format!("{}{}\n", indent, name));
                        } else {
                            output.push_str(&format!("{}{}\n", indent, name));
                            for sig in &sigs {
                                output.push_str(&format!("{}  {}\n", indent, sig));
                            }
                        }
                    } else {
                        output.push_str(&format!("{}{}\n", indent, name));
                    }
                }
            }
        }
        Ok(())
    }

    walk(
        root, root, pattern, detail, &mut output, &mut file_count, &mut dir_count, 0,
    )?;

    output.push_str(&format!(
        "\n{} directories, {} files",
        dir_count, file_count
    ));

    Ok(output)
}

fn matches_pattern(path: &str, pattern: &str) -> bool {
    if pattern == "**/*" || pattern == "*" {
        return true;
    }
    // Simple glob matching
    if pattern.starts_with("**/*.") {
        let ext = &pattern[5..];
        return path.ends_with(ext);
    }
    if pattern.starts_with("*.") {
        let ext = &pattern[2..];
        return path.rsplit('/').next().map(|n| n.ends_with(ext)).unwrap_or(false);
    }
    path.contains(pattern)
}

/// Extract function/class signatures from a source file.
fn extract_signatures(path: &Path) -> Vec<String> {
    let ext = path.extension().and_then(|e| e.to_str()).unwrap_or("");
    let content = std::fs::read_to_string(path).unwrap_or_default();

    match ext {
        "py" => extract_python_signatures(&content),
        "rs" => extract_rust_signatures(&content),
        "ts" | "tsx" | "js" | "jsx" => extract_js_signatures(&content),
        _ => Vec::new(),
    }
}

fn extract_python_signatures(content: &str) -> Vec<String> {
    let mut sigs = Vec::new();
    for line in content.lines() {
        let trimmed = line.trim();
        if trimmed.starts_with("def ") || trimmed.starts_with("class ") || trimmed.starts_with("async def ") {
            if let Some(end) = trimmed.find(':') {
                sigs.push(trimmed[..end].to_string());
            }
        }
    }
    sigs
}

fn extract_rust_signatures(content: &str) -> Vec<String> {
    let mut sigs = Vec::new();
    for line in content.lines() {
        let trimmed = line.trim();
        if (trimmed.starts_with("pub fn ") || trimmed.starts_with("fn ")
            || trimmed.starts_with("pub async fn ") || trimmed.starts_with("async fn ")
            || trimmed.starts_with("pub struct ") || trimmed.starts_with("struct ")
            || trimmed.starts_with("pub enum ") || trimmed.starts_with("enum ")
            || trimmed.starts_with("pub trait ") || trimmed.starts_with("trait ")
            || trimmed.starts_with("impl "))
            && !trimmed.starts_with("//")
        {
            // Take up to the opening brace or semicolon
            let sig = if let Some(pos) = trimmed.find('{') {
                trimmed[..pos].trim_end().to_string()
            } else if let Some(pos) = trimmed.find(';') {
                trimmed[..pos].trim_end().to_string()
            } else {
                trimmed.to_string()
            };
            sigs.push(sig);
        }
    }
    sigs
}

fn extract_js_signatures(content: &str) -> Vec<String> {
    let mut sigs = Vec::new();
    for line in content.lines() {
        let trimmed = line.trim();
        if (trimmed.starts_with("export function ") || trimmed.starts_with("function ")
            || trimmed.starts_with("export async function ") || trimmed.starts_with("async function ")
            || trimmed.starts_with("export const ") || trimmed.starts_with("const ")
            || trimmed.starts_with("export class ") || trimmed.starts_with("class "))
            && !trimmed.starts_with("//")
        {
            let sig = if let Some(pos) = trimmed.find('{') {
                trimmed[..pos].trim_end().to_string()
            } else {
                trimmed.to_string()
            };
            sigs.push(sig);
        }
    }
    sigs
}
